### ISSUE TITLE
Naming improvements for the Dhall module

### ISSUE THREAD

**Issue by sjakobi** [dhall-lang/dhall-haskell#1393]
@Gabriel439 recently mentioned that `Interpret` should be renamed to `FromDhall`, and `Inject` to `ToDhall`. That makes a lot of sense to me.

IMHO there are more things in the `Dhall` module that deserve better names:

* `Type` and `InputType`.
* `input` suggests a link to `InputType` but is actually more related to `Interpret`.
* `InterpretOptions` are used both by `Interpret` and by `Inject`.

This renaming would be a large breaking change, so we should try find a naming scheme that we're happy with and then implement it within a single version bump.

Here's what I perceive to be the "big" names and exports from `Dhall`:

### Dhall to Haskell

```haskell
class Interpret a where
    autoWith :: InterpretOptions -> Type a

auto :: Interpret a => Type a

data Type a = Type
    { extract  :: Expr Src X -> Extractor Src X a
    , expected :: Expr Src X
    }

input :: Type a -> Text -> IO a
```

### Haskell to Dhall

```haskell
class Inject a where
    injectWith :: InterpretOptions -> InputType a

inject :: Inject a => InputType a

data InputType a = InputType
    { embed    :: a -> Expr Src X
    , declared :: Expr Src X
    }
```

**Comment by sjakobi** [comment]
Here's an initial idea for a better naming scheme:

### Dhall to Haskell

```haskell
class FromDhall a where
    decoderWith :: MarshalOptions -> Decoder a

defaultDecoder :: FromDhall a => Decoder a

data Decoder a = Decoder
    { decode   :: Expr Src X -> Extractor Src X a -- or keep 'extract'?
    , expected :: Expr Src X
    }

decodeIO :: Decoder a -> Text -> IO a
```

### Haskell to Dhall

```haskell
class ToDhall a where
    encoderWith :: MarshalOptions -> Encoder a

defaultEncoder :: ToDhall a => Encoder a

data Encoder a = Encoder
    { encode   :: a -> Expr Src X
    , declared :: Expr Src X
    }
```

**Comment by Gabriel439** [comment]
@sjakobi: We don't need to rename everything at once.  In fact, I think it's actually better if we spread breaking changes over time as I discuss here:

* [Release early and often](http://www.haskellforall.com/2019/05/release-early-and-often.html)

I think the first step should be just renaming the classes to `FromDhall` and `ToDhall`

**Comment by sjakobi** [comment]
I have updated the issue description with the changes from #1437. What should be the next step?

I still like the idea of renaming `Type` on the `FromDhall` side to `Decoder` and `InputType` on the `ToDhall` side to `Encoder`.

**Comment by Gabriel439** [comment]
@sjakobi: Yeah, that seems reasonable to me, although we might want to postpone that second breaking change to the next release

**Comment by sjakobi** [comment]
> that second breaking change

Are you referring to both `Type -> Decoder` and `InputType -> Encoder` here or just to `InputType -> Encoder`? I.e. would it be ok to include both in the upcoming v1.28?

**Comment by Gabriel439** [comment]
@sjakobi: It's fine to bundle both of those changes in the same release
